universe/Building.cpp

```diff
@@ -239,13 +239,18 @@ float BuildingType::ProductionCost(int empire_id, int location_id) const {
         if (m_production_cost && m_production_cost->ConstantExpr())
             return m_production_cost->Eval();
 
+        const auto arbitrary_large_number = 999999.9f;
+
         std::shared_ptr<UniverseObject> location = GetUniverseObject(location_id);
         if (!location)
-            return 999999.9f;    // arbitrary large number
+            return arbitrary_large_number;
 
         std::shared_ptr<const UniverseObject> source = Empires().GetSource(empire_id);
-        if (!source || m_production_cost->SourceInvariant())
-            return 999999.9f;
+        if (!source && !m_production_cost->SourceInvariant()) {
+            ErrorLogger() << "Source is expected but missing using production cost of "
```

Comment:
same here... not always an error / unexpected.

the message is also poorly worded / ambiguous